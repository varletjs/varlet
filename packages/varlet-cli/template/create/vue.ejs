<template>
  <div class="var-<%- kebabCaseName %>"></div>
</template>

<script lang="ts">
import { defineComponent } from 'vue'
import './<%- kebabCaseName %>.less'

export default defineComponent({
  name: '<%- namespace + bigCamelizeName %>',
})
</script>
