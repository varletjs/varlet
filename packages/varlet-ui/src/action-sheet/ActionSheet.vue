<template>
  <var-popup
    :class="n('popup-radius')"
    position="bottom"
    :overlay="overlay"
    :overlay-class="overlayClass"
    :overlay-style="overlayStyle"
    :lock-scroll="lockScroll"
    :close-on-click-overlay="closeOnClickOverlay"
    :teleport="teleport"
    :show="popupShow"
    v-bind="{
      'onUpdate:show': handlePopupUpdateShow,
    }"
    @open="onOpen"
    @close="onClose"
    @closed="onClosed"
    @opened="onOpened"
    @route-change="onRouteChange"
  >
    <div :class="classes(n(), 'var--box')" v-bind="$attrs">
      <slot name="title">
        <div :class="n('title')">{{ dt(title, pack.actionSheetTitle) }}</div>
      </slot>

      <slot name="actions">
        <div
          :class="classes(n('action-item'), action.className, [action.disabled, n('--disabled')])"
          v-ripple="{ disabled: action.disabled }"
          v-for="action in actions"
          :key="action.name"
          :style="{ color: action.color }"
          @click="handleSelect(action)"
        >
          <var-icon
            :class="n('action-icon')"
            var-action-sheet-cover
            :name="action.icon"
            :size="action.iconSize"
            v-if="action.icon"
          />
          <div :class="n('action-name')">{{ action.name }}</div>
        </div>
      </slot>
    </div>
  </var-popup>
</template>

<script lang="ts">
import Ripple from '../ripple'
import VarPopup from '../popup'
import VarIcon from '../icon'
import { defineComponent, ref, watch } from 'vue'
import { props } from './props'
import { dt } from '../utils/shared'
import { pack } from '../locale'
import { createNamespace, call } from '../utils/components'
import type { Ref } from 'vue'
import type { ActionItem } from './index'

const { n, classes } = createNamespace('action-sheet')

export default defineComponent({
  name: 'VarActionSheet',
  directives: { Ripple },
  components: {
    VarPopup,
    VarIcon,
  },
  inheritAttrs: false,
  props,
  setup(props) {
    const popupShow: Ref<boolean> = ref(false)

    const handleSelect = (action: ActionItem) => {
      if (action.disabled) {
        return
      }

      const { closeOnClickAction, onSelect } = props
      call(onSelect, action)
      closeOnClickAction && call(props['onUpdate:show'], false)
    }

    const handlePopupUpdateShow = (value: boolean) => call(props['onUpdate:show'], value)

    watch(
      () => props.show,
      (newValue) => {
        popupShow.value = newValue
      },
      { immediate: true }
    )

    return {
      n,
      classes,
      handlePopupUpdateShow,
      popupShow,
      pack,
      dt,
      handleSelect,
    }
  },
})
</script>

<style lang="less">
@import '../styles/common';
@import '../icon/icon';
@import '../ripple/ripple';
@import '../popup/popup';
@import './actionSheet';
</style>
