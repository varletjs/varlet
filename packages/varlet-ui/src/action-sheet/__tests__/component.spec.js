import ActionSheet from '../index'
import VarActionSheet from '../ActionSheet'
import { mount } from '@vue/test-utils'
import { createApp } from 'vue'

test('test action-sheet component plugin', () => {
  const app = createApp({}).use(ActionSheet.Component)
  expect(app.component(ActionSheet.Component.name)).toBeTruthy()
})

test('test action-sheet styles', () => {
  const Wrapper = {
    components: {
      [VarActionSheet.name]: VarActionSheet,
    },
    data: () => ({
      actions: [
        {
          name: 'Item 01',
          icon: 'account-circle',
          iconSize: '30px',
          color: '#4caf50',
          className: 'customClass',
        },
        {
          name: 'Item 02',
          icon: 'account-circle',
          disabled: true,
        },
      ],
    }),
    template: `
      <var-action-sheet :actions="actions" />
    `,
  }
  const wrapper = mount(Wrapper)
  expect(wrapper.html()).toMatchSnapshot()
  wrapper.unmount()
})
