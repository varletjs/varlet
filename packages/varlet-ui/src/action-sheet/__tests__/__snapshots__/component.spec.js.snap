// Vitest Snapshot v1

exports[`test action-sheet styles 1`] = `
"<transition-stub>
  <div class=\\"var--box var-popup\\" style=\\"z-index: 1998; display: none;\\">
    <div class=\\"var-popup__overlay\\" style=\\"z-index: 1999;\\"></div>
    <transition-stub>
      <!---->
    </transition-stub>
  </div>
</transition-stub>"
`;
